crates/fortitude_preprocessor/src/cpp.rs

```diff
@@ -0,0 +1,852 @@
+    SystemDefined,
+    /// A user-defined macro, e.g. passed in via command line.
+    UserDefined,
+    /// A macro defined in the source file.
+    FileDefined {
+        start: usize,
+        end: usize,
+        path: PathBuf,
+    },
+    /// Plain text from an included file.
+    IncludeText {
+        start: usize,
+        end: usize,
+        path: PathBuf,
+    },
+    /// Plain text in the source file, not from a macro.
+    LocalText { start: usize, end: usize },
+}
+
+/// A snippet of text with provenance information.
+pub struct Snippet {
+    text: String,
+    provenance: Provenance,
+}
+
+impl Snippet {
+    pub fn extend(&mut self, other: &Snippet) -> anyhow::Result<()> {
+        match (&self.provenance, &other.provenance) {
+            (
+                Provenance::LocalText {
+                    start: self_start,
+                    end: self_end,
+                },
+                Provenance::LocalText {
+                    start: other_start,
+                    end: other_end,
+                },
+            ) => {
+                if self_end == other_start {
+                    self.text.push_str(&other.text);
+                    self.provenance = Provenance::LocalText {
+                        start: *self_start,
+                        end: *other_end,
+                    };
+                    Ok(())
+                } else {
+                    Err(anyhow!(
+                        "Can only extend snippets with contiguous provenance"
+                    ))
+                }
+            }
+            _ => Err(anyhow!(
+                "Can only extend snippets with LocalText provenance"
+            )),
+        }
+    }
+}
+
+/// A collection of snippets.
+pub struct Snippets {
+    // Collected snippets
+    inner: Vec<Snippet>,
+}
+
+impl Default for Snippets {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Snippets {
+    pub fn new() -> Self {
+        Snippets { inner: Vec::new() }
+    }
+
+    pub fn push(&mut self, text: &str, provenance: &Provenance) {
+        let snippet = Snippet {
+            text: text.to_string(),
+            provenance: provenance.clone(),
+        };
+        if let Some(last) = self.inner.last_mut() {
+            if last.extend(&snippet).is_ok() {
+                return;
+            }
+        }
+        self.inner.push(snippet);
+    }
+
+    pub fn collect(&self) -> String {
+        self.inner
+            .iter()
+            .map(|s| s.text.as_str())
+            .collect::<String>()
+    }
+}
+
+/// A logical line of code, which may span multiple physical lines due to
+/// line continuations. Tracks the byte offset of each location of the
+/// logical line.
+pub struct LogicalLine {
+    /// The text of the logical line.
+    text: String,
+    /// The byte offsets of each character in the logical line
+    /// relative to the start of the source file.
+    byte_offsets: Vec<usize>,
```

Comment:
Storing offset of each character individually worries me a little bit. Potential optimisation would be `Vec<TextRange>` storing contiguous chunks?

Reply:
Yes, I was planning to optimise that at some point. I was thinking it would only be necessary to store the discontinuities, but thought I would implement it the easy way to begin.

Reply:
:+1: Sounds good! Make it work first, then make it fast :)